WalletWasabi/WabiSabi/Backend/DoSPrevention/Prison.cs

```diff
@@ -110,7 +111,9 @@ TimeFrame CalculatePunishmentInheritance(OutPoint[] ancestors)
 				return cachedBanningTime;
 			}
 
-			offender = Offenders.LastOrDefault(x => x.OutPoint == outpoint);
```

Comment:
Yes this definitely looks like the reason of poor backend performances. I believe changing that will improve a lot situation, but maybe it would be good to add logging for benchmarks of Prison related functions.